ext/memcached_rb/memcached_rb.c

```diff
@@ -1,460 +1,518 @@
-  return rb_obj;
+	memcached_st *mc = memcached_create(NULL);
+	return Data_Wrap_Struct(klass, NULL, memcached_free, mc);
 }
 
 static VALUE
-rb_connection_initialize(VALUE self, VALUE rb_servers)
+rb_connection_new(VALUE klass, VALUE rb_servers)
 {
-  memcached_ctx *ctx = get_ctx(self);
-  int i;
-  memcached_return_t rc;
-
-  for (i = 0; i < RARRAY_LEN(rb_servers); ++i) {
-    VALUE rb_server = rb_ary_entry(rb_servers, i);
-
-    VALUE rb_backend = rb_ary_entry(rb_server, 0);
-
-    if (id_tcp == SYM2ID(rb_backend)) {
-      VALUE rb_hostname = rb_ary_entry(rb_server, 1);
-      VALUE rb_port = rb_ary_entry(rb_server, 2);
-      // TODO: add weight
-      //VALUE rb_weight = rb_ary_entry(rb_server, 3);
-      rc = memcached_server_add (ctx->memc, StringValueCStr(rb_hostname), NUM2INT(rb_port));
-      handle_memcached_return(rc);
-    } else if (id_socket == SYM2ID(rb_backend)) {
-      VALUE rb_fd = rb_ary_entry(rb_server, 1);
-      rc = memcached_server_add_unix_socket (ctx->memc, StringValueCStr(rb_fd));
-      handle_memcached_return(rc);
-    }
-  }
-
-  // Verify keys by default
-  rc = memcached_behavior_set(ctx->memc, MEMCACHED_BEHAVIOR_VERIFY_KEY, true);
-  handle_memcached_return(rc);
-
-  return Qnil;
+	VALUE rb_mc = rb_memcached_alloc(klass);
+
+	memcached_st *mc;
+	memcached_return_t rc;
+	long i;
+
+	UnwrapMemcached(rb_mc, mc);
+	Check_Type(rb_servers, T_ARRAY);
+
+	for (i = 0; i < RARRAY_LEN(rb_servers); ++i) {
+		VALUE rb_server, rb_backend;
+		
+		rb_server = rb_ary_entry(rb_servers, i);
+		Check_Type(rb_server, T_ARRAY);
+
+		rb_backend = rb_ary_entry(rb_server, 0);
+		Check_Type(rb_backend, T_SYMBOL);
+
+		if (SYM2ID(rb_backend) == id_tcp) {
+			VALUE rb_hostname = rb_ary_entry(rb_server, 1);
+			VALUE rb_port = rb_ary_entry(rb_server, 2);
+
+			Check_Type(rb_server, T_STRING);
+			Check_Type(rb_port, T_FIXNUM);
+
+			// TODO: add weight
+			//VALUE rb_weight = rb_ary_entry(rb_server, 3);
+			rc = memcached_server_add(mc, StringValueCStr(rb_hostname), NUM2INT(rb_port));
+			rb_memcached_error_check(rc);
+		}
+		else if (SYM2ID(rb_backend) == id_socket) {
+			VALUE rb_fd = rb_ary_entry(rb_server, 1);
+			Check_Type(rb_fd, T_STRING);
+			rc = memcached_server_add_unix_socket(mc, StringValueCStr(rb_fd));
+			rb_memcached_error_check(rc);
+		}
+	}
+
+	// Verify keys by default
+	rc = memcached_behavior_set(mc, MEMCACHED_BEHAVIOR_VERIFY_KEY, true);
+	rb_memcached_error_check(rc);
+
+	return rb_mc;
 }
 
 static memcached_return_t
 iterate_server_function(const memcached_st *ptr, const memcached_instance_st * server, void *context)
 {
-  VALUE server_list = (VALUE) context;
-
-  VALUE rb_server = rb_obj_alloc(rb_cServer);
-
-  rb_ivar_set(rb_server, id_ivar_hostname, rb_str_new2(memcached_server_name(server)));
-  rb_ivar_set(rb_server, id_ivar_port, UINT2NUM(memcached_server_port(server)));
-  // TODO weight
+	VALUE rb_server_list = (VALUE)context;
+	VALUE rb_server = rb_funcall(rb_cServer, rb_intern("new"), 2, 
```

Comment:
Another big change worth pointing out: there was a lot of boilerplate code around here because the `Server` class was being defined in C. Too hard! You can define it in 5 lines of Ruby (see `lib/memcached/server.rb`), and then simply call the constructor from C to create servers.